### PR TITLE
Signer module - `kairos-crypto`

### PR DESCRIPTION
**Pull request by koxu1996**
This PR moves [cryptography](https://github.com/cspr-rad/kairos/tree/4309cd11d55b900fe4fd8982ee9ea763008618d3/kairos-cli/src/crypto) from `kairos-cli` - introduced temporarily in [Base CLI parsing #12](https://github.com/cspr-rad/kairos/pull/17) - into a separate package named `kairos-crypto`.

There are few major improvements, mainly:

1. Introduced `CryptoSigner` trait to provide **implementation-agnostic interface**.
2. Casper specific implementation is available via feature flag `crypto-casper` (enabled by default).
3. Introduced `sign()` and `verify()` methods - it will be used for signature checking, once https://github.com/cspr-rad/kairos/pull/30 is merged.
4. Serialization error was split into distinct `CryptoError::Serialization` and `CryptoError::Deserialization`.

### CODE REVIEW

**Comment by koxu1996** [comment]
Rebased and ready for review.

File: kairos-crypto/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+pub trait CryptoSigner {
+    fn from_private_key_file<P: AsRef<Path>>(file: P) -> Result<Self, CryptoError>
+    where
+        Self: Sized;
+    fn from_public_key(bytes: &[u8]) -> Result<Self, CryptoError>
+    where
+        Self: Sized;
+
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>, CryptoError>;
+    fn verify(&self, data: &[u8], signature_bytes: &[u8]) -> Result<(), CryptoError>;
```
Comment (Avi-D-coder) [inline kairos-crypto/src/lib.rs:17]: I'd prefer `AsRef<[u8]>` instead of `&[u8]`.

**Review by Avi-D-coder — APPROVED** [review]
This is fine.
I don't love traits with a single impl, but since we are planing on adding more it's good to merge.

File: kairos-crypto/src/implementations/casper.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            public_key,
+        })
+    }
+
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>, CryptoError> {
+        let private_key = self
+            .private_key
+            .as_ref()
+            .ok_or(CryptoError::MissingPrivateKey)?;
+        let signature = crypto::sign(data, private_key, &self.public_key);
```
Comment (xcthulhu) [inline kairos-crypto/src/implementations/casper.rs:52]: Mentioned elsewhere, `casper_types::crypto::sign` has an unused `public_key`.

Here is the crypto signing code that `litmus` uses (which has been unit tested and exhaustively tested against `casper_types` for compatibility).

https://github.com/cspr-rad/litmus/blob/673f8182560f7f1ff6fee04d775070a6b3e380c3/src/crypto.rs#L24-L40

```rust
// Signs the given message using the given key.
pub fn sign<T: AsRef<[u8]>>(secret_key: &SecretKey, message: T) -> Signature {
    match secret_key {
        SecretKey::System => Signature::System,
        SecretKey::Ed25519(secret_key) => {
            let signature = secret_key.sign(message.as_ref());
            Signature::Ed25519(signature)
        }
        SecretKey::Secp256k1(secret_key) => {
            let signature = secret_key
                .try_sign(message.as_ref())
                .expect("should create signature");
            Signature::Secp256k1(signature)
        }
        _ => panic!("SecretKey is marked as non-exhaustive, but this should never happen"),
    }
}
```

File: kairos-crypto/src/lib.rs (hunk 1 above)
Comment (koxu1996) [reply kairos-crypto/src/lib.rs]: Fixed in https://github.com/cspr-rad/kairos/pull/35/commits/9792ee33cc357f3798cca652a76e89cc4ff51755.

File: kairos-crypto/src/implementations/casper.rs (hunk 2 above)
Comment (koxu1996) [reply kairos-crypto/src/implementations/casper.rs:50]: I am aware that public key is not used internally, but I see no point in duplicating the Casper implementation just to get rid of it:

image

Additionally, public key is always available in `Signer`, so there is no extra code necessary.

**Review by Avi-D-coder — APPROVED** [review]

**Review by Rom3dius — APPROVED** [review]
Looks good to go

**Review by Avi-D-coder — APPROVED** [review]

**Pull request merged**